Refactor TransformStream's backpressure handling

tyoshino opened the pull request:
Add notes (in the form of assert description where possible) about the new
backpressure controlling logic introduced by isonmad to ease understanding.

tyoshino commented:
/cc @isonmad

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -97,40 +105,51 @@ function TransformStreamErrorInternal(transformStream, e) {
-  assert(transformStream._backpressureChangePromise !== undefined);
+// Used for preventing the next write() call on TransformStreamSink until there
+// is no longer backpressure.
+function TransformStreamBackpressureGonePromise(transformStream) {
```
I don't see how this name change makes it clearer. ReadyPromise is analogous to the [[readyPromise]] that writablestream writers have.

tyoshino replied:
OK. How about ReadableReadyPromise?

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -363,14 +386,18 @@ module.exports = class TransformStream {
 
     const desiredSize = this._readableController.desiredSize;
+    // Tentatively set _backpressure based on desiredSize. When desiredSize is 0, it's possible that a pull() is made
+    // immediately after this because of pending read()s and fix _backpressure back to false.
```
There can't be pending reads because this is the constructor, there's nothing 'tentative' about the fact that there is or isn't backpressure when the stream is first constructed.

tyoshino replied:
Oops, yeah, this is not right for here. Fixed.

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -363,14 +386,18 @@ module.exports = class TransformStream {
-    const sink = new TransformStreamSink(this, startPromise);
+    const sink = new TransformStreamSink(this, startPromise.then(() => {
+      return TransformStreamBackpressureGonePromise(this);
+    }));
```
I don't see the point of doing this in the constructor instead of inside sink.start. It's conceptually something that .start does, and it's symmetrical with what source.start does. It's not really any clearer this way.

tyoshino replied:
OK. Reverted

tyoshino replied:
I felt that it's more readable to me to place logic exchanging signal between sink and source or doing something based on stuff stored inside TransformStream. TransformStream is already controlling the completion of sink.start by passing startPromise. I was viewing this as an additional delaying logic from TransformStream. And, this signal is based on observation on the source. That's my thought behind this suggestion.

But your points also makes sense. So, I'll go with your original structure.

isonmad replied:
The change isn't fully reverted; the original explanatory comment is gone, and the code didn't need to be in a block per [ricea's review at the original PR](https://github.com/whatwg/streams/pull/550#discussion_r84678242).

tyoshino replied:
Oh, sorry! Addressing in #577

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -238,13 +256,17 @@ class TransformStreamSource {
+      // Prevent the first pull() call until there is backpressure.
+
+      assert(transformStream._backpressureChangePromise !== undefined,
+             '_backpressureChangePromise should have been initialized');
```
This added assert is good.

tyoshino commented:
> I still don't see how this improves code comprehensibility, honestly.

I've addressed your points. Is there anything else doesn't make sense to you?

ricea commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -45,10 +47,16 @@ function TransformStreamEnqueueToReadable(transformStream, chunk) {
+    // This allows pull() again. When desiredSize is 0, it's possible that a pull() is made immediately (but
+    // asynchronously) after this because of pending read()s and fix _backpressure back to false.
+    //
+    // We don't have to worry about that such a pull() may happen synchronously to the enqueue(), and therefore is not
```
Sorry, I don't understand what "synchronously to" means here. Maybe you could explain the dangerous case as a series of steps, and then explain why it never happens?

ricea commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -97,40 +105,51 @@ function TransformStreamErrorInternal(transformStream, e) {
   // console.log(`TransformStreamSetBackpressure(${backpressure})`);
 
-  assert(transformStream._readableBackpressure !== backpressure);
+  // Passes also when called for initialization.
```
How about "during construction" instead of "for initialization"?

ricea commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -97,40 +105,51 @@ function TransformStreamErrorInternal(transformStream, e) {
+         'TransformStreamSetBackpressure() should be called only when backpressure is changed');
 
   if (transformStream._backpressureChangePromise !== undefined) {
+    // The fulfillment value is just for the assertion.
```
"the assertion" is ambiguous because it's not clear which assertion is being referred to. How about "just for a sanity check"?

tyoshino replied:
Done

tyoshino commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -45,10 +47,16 @@ function TransformStreamEnqueueToReadable(transformStream, chunk) {
-    TransformStreamSetBackpressure(transformStream, backpressure);
+  if (maybeBackpressure === true && transformStream._backpressure === false) {
+    // This allows pull() again. When desiredSize is 0, it's possible that a pull() is made immediately (but
+    // asynchronously) after this because of pending read()s and fix _backpressure back to false.
```
Done

tyoshino commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -45,10 +47,16 @@ function TransformStreamEnqueueToReadable(transformStream, chunk) {
-  if (transformStream._readableBackpressure !== backpressure) {
-    TransformStreamSetBackpressure(transformStream, backpressure);
+  if (maybeBackpressure === true && transformStream._backpressure === false) {
+    // This allows pull() again. When desiredSize is 0, it's possible that a pull() is made immediately (but
```
Done

ricea commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -45,10 +47,16 @@ function TransformStreamEnqueueToReadable(transformStream, chunk) {
+    // This allows pull() again. When desiredSize is 0, it's possible that a pull() will happen immediately (but
+    // asynchronously) after this because of pending read()s and set _backpressure back to false.
+    //
+    // We don't have to worry about that such a pull() may happen inside the enqueue() call, and therefore is not
```
How about:

> If pull() could be called from inside enqueue(), then this logic would be wrong. This cannot happen because there is always a promise pending from start() or pull() when _backpressure is false.

tyoshino replied:
Thanks for the nice rewriting. Done

ricea commented:
lgtm

tyoshino commented:
Thank you isonmad, ricea for review.

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -371,6 +395,8 @@ module.exports = class TransformStream {
     assert(this._readableController !== undefined);
 
     const desiredSize = this._readableController.desiredSize;
+    // Set _backpressure based on desiredSize. As there is no read() at this point, so we can just interpret
```
"As there is no read()at this point, we can"

tyoshino replied:
Going to address in #577

isonmad commented on reference-implementation/lib/transform-stream.js:
```diff
@@ -45,10 +47,15 @@ function TransformStreamEnqueueToReadable(transformStream, chunk) {
+    // asynchronously) after this because of pending read()s and set _backpressure back to false.
+    //
+    // If pull() could be called from inside enqueue(), then this logic would be wrong. This cannot happen
+    // because there is always a promise pending from start() or pull() when _backpressure is false.
```
Okay, with these two lines now I finally understand what you were worried about this entire time. I really had no idea what you were talking about.

tyoshino replied:
Sorry for not being clear enough. Thanks for review.
